src/iframe-helper.js

```diff
@@ -380,3 +380,61 @@ function parseIfNeeded(data) {
   }
   return data;
 }
+
+
+
+/**
+ * Manages a postMessage API for an iframe with a subscription message and
+ * a way to broadcast messages to all subscribed windows, which
+ * in turn must all be descendants of the contentWindow of the iframe.
+ */
+export class SubscriptionApi {
```

Comment:
I am simply not convinced what performance gain do we get by rewriting it to:
```
/**
  * @param {!Element} iframe The iframe.
  * @param {string} type Type of the subscription message.
  * @param {boolean} is3p set to true if the iframe is 3p.
  * @param {function(!Object, !Window, string)} requestCallback Callback
  *     invoked whenever a new window subscribes.
  * @return {function(string, !Object)} Function which sends a message of the provided type and
  *     with the provided payload to all subscribed windows.
  */
export function registerSubscriptionApi(iframe, type, is3p, requestCallback) {
  var clientWindows = [];
  listenFor(
      iframe, type, (data, source, origin) => {
        // This message might be from any window within the iframe, we need
        // to keep track of which windows want to be sent updates.
        if (!clientWindows.some(entry => entry.win == source)) {
          clientWindows.push({win: source, origin});
        }
        requestCallback(data, source, origin);
      }, is3p,
      // For 3P frames we also allow nested frames within them to subscribe..
      is3p /* opt_includingNestedWindows */);
  return (type, data) => postMessageToWindows(iframe, clientWindows, type, data, is3p);
}
```

Sure, it looks like less code due to formatting. However, the following issues exist:

- to someone reading the code it is not immediately obvious that there is state involved which might lead to bugs / memory leaks in the future
- memory allocation should be about the same, we're returning a function closure which will need to be heap-allocated anyway, just like an instance of the class with all the same data
- this code is executed once per instantiation of IntersectionObserver, if we're worried about the nested call to init_ from the constructor then we could inline that but it feels like premature optimisation
- we lose the ability to ever unregister the listener we set up without reverting back to returning an object
- we need to jump through more hoops in the tests since we can no longer access clientWindows, though I can't tell whether this will look better or worse until I actually do it

AFAIK, modern JS engines are very good at handling class-like constructs, so I'd expect instantiation of that class to be of a similar cost to allocating a closure on the heap. I have to assume I am missing something here with respect to the performance characteristics of this code and so I will make the requested change, but I would definitely appreciate it if you could tell me what it is that I am missing:)